src/ctypes-foreign-threaded/foreign.mli

```diff
@@ -93,3 +104,88 @@ val funptr_opt :
 exception CallToExpiredClosure
 (** A closure passed to C was collected by the OCaml garbage collector before
     it was called. *)
+
+module type Funptr = sig
+  type fn
+  (** [fn] is the signature of the underlying ocaml function. *)
+
+  type t
+  (** Handle to an ocaml function that can be passed to C for use in callbacks.
+
+      Unfortunately it is not possible to track if a fuction pointer is still used in C,
+      so you must use the appropriate life cycle functions below for this.
+
+      See [t], [of_fun], [with_fun] and [free] *)
+
+  val t : t Ctypes.typ
+  (** ctype that can be used in type signatures of external functions *)
+
+  val t_opt : t option Ctypes.typ
+  (** Like [t] but optional. *)
+
+  val free : t -> unit
+  (** [free fptr] - Indicate that the [fptr] is no longer needed.
+
+      Once [free] has been called any C calls to this [Dynamic_funptr.t] are
+      unsafe and may result in a segmentation fault. Only call [free] once the
+      callback is no longer used from C.
+  *)
+
+  val of_fun : ?debug_info:string -> fn -> t
+  (** [of_fun fn] - Turn an ocaml closure into a function pointer
+      that can be passed to C.
+
+      You MUST call [free] to the function pointer is no longer needed.
+      Failure to do so will result in a memory leak.
+      Failure to call [free] and not holding a reference this this pointer
+      is an error. To avoid unexpected crashes we log this and ensure that
+      potentially still needed ocaml values are retained.
+
+      For many use cases it may be simpler to use [with_fun] instead as this will
+      do the free for you. *)
+
+  val with_fun : ?debug_info:string -> fn -> (t -> 'c) -> 'c
+  (** [with_fun fn (fun fptr -> DO_STUFF)] - Turn an ocaml closure into a
+      function pointer and do simple life cycle management.
+
+      This will automatically call [free fptr] after [DO_STUFF] completes.
+
+      It is not safe to use if the C function ptr [fptr] may still be used
+      after [DO_STUFF] completed.
+  *)
+end
+
+(** [(val (dynamic_funptr fn))] - define a Ctype for more safely passing ocaml
+    functions to C.
+
+    [(val (funptr_spec (FOO @-> returning BAR)))] is roughly equivalent to
+    [BAR( * )(FOO)] in C.
```

Comment:
I think this comment (about `funptr_spec`) is no longer needed.

Reply:
comment has been updated.